ModelViewer/ModelViewer/ModelLoader.cpp

```diff
@@ -25,154 +26,178 @@ Model ModelLoader::LoadModel(const QString& file)
-		}
-
-
-		// Load indices
-		newMesh.m_indexCount = pMesh->mNumFaces * 3;
-		const int indexBufferSize = sizeof(int) * newMesh.m_indexCount;
-		newMesh.m_indexBuffer.allocate(indexBufferSize);
-
-		std::vector<int> indices;
-		indices.reserve(newMesh.m_indexCount);
-
-		// Loop through all faces
-		// Potentially use OpenMP to multithread this if it is slow
-		for (int j = 0; j < pMesh->mNumFaces; ++j) {
-			aiFace const& face = pMesh->mFaces[j];
-			indices.push_back(face.mIndices[0]);
-			indices.push_back(face.mIndices[1]);
-			indices.push_back(face.mIndices[2]);
-		}
-
-		newMesh.m_indexBuffer.write(0, indices.data(), indexBufferSize);
-
-
-		// Load axis aligned bounding box (AABB)
-		newMesh.m_AABBMin.setX(pMesh->mAABB.mMin.x);
-		newMesh.m_AABBMin.setY(pMesh->mAABB.mMin.y);
-		newMesh.m_AABBMin.setZ(pMesh->mAABB.mMin.z);
-		newMesh.m_AABBMax.setX(pMesh->mAABB.mMax.x);
-		newMesh.m_AABBMax.setY(pMesh->mAABB.mMax.y);
-		newMesh.m_AABBMax.setZ(pMesh->mAABB.mMax.z);
-
-
-		// Append this mesh to the list
-		ret.m_meshes.push_back(newMesh);
+	// Get pointers to the important information from ASSIMP
+	aiMesh const* pMesh = pScene->mMeshes[meshIdx];
+	aiMaterial const* pMaterial = pScene->mMaterials[pMesh->mMaterialIndex];
+
+
+	auto AssignColor = [](GLfloat* vec, aiColor3D col) {
+		vec[0] = col.r;
+		vec[1] = col.g;
+		vec[2] = col.b;
+		vec[3] = 1.f; // No alpha
+	};
+
+	// Set up the material for this mesh
+	aiColor3D color;
+	float val;
+	pMaterial->Get(AI_MATKEY_COLOR_AMBIENT, color);
+	AssignColor(newMesh.m_ambient, color);
+	pMaterial->Get(AI_MATKEY_COLOR_DIFFUSE, color);
+	AssignColor(newMesh.m_diffuse, color);
+	pMaterial->Get(AI_MATKEY_COLOR_SPECULAR, color);
+	AssignColor(newMesh.m_specular, color);
+	pMaterial->Get(AI_MATKEY_SHININESS, val);
+	newMesh.m_shininess = val;
+
+
+
+	// TODO: Handle textures
+
+
+
+	// Determine if we have colors, normals, or texture coordinates
+	newMesh.m_hasNormals = (pMesh->mNormals != NULL);
+	newMesh.m_hasUVCoordinates = pMesh->HasTextureCoords(0);
+	newMesh.m_hasColors = pMesh->HasVertexColors(0);
+
+	// Determine the size of a single attribute
+	newMesh.m_numPositionComponents = 3;
+	newMesh.m_numNormalComponents = 3;
+	newMesh.m_numUVComponents = pMesh->mNumUVComponents[0];  // Various
+	newMesh.m_numColorComponents = 4;
+
+	// Compute sizes for all attributes
+	const int positionBufferSize = newMesh.m_numPositionComponents * pMesh->mNumVertices * sizeof(float);
+	const int normalBufferSize = newMesh.m_numNormalComponents * pMesh->mNumVertices * sizeof(float);
+	const int uvBufferSize = newMesh.m_numUVComponents * pMesh->mNumVertices * sizeof(float);
+	const int colorBufferSize = newMesh.m_numColorComponents * pMesh->mNumVertices * sizeof(float);
+
+	// Compute the offsets of features in the overall vertex buffer
+	//   1. Positions (required)
+	//   2. Normals (optional)
+	//   3. UV Coordinates (optional)
+	//   4. Colors (optional)
+	int currentOffset = 0;
+	if (true) {
```

Comment:
If (true)? Is this intentional so it matches the form of the if statements below?

Reply:
This section was copied from one function to another... I didn't re-write it.  I think the idea was to match the format below.  `if(true)` because there will always be vertex locations.  I'll add a comment though :)